### PR TITLE
Mx 14495 inject run type components

### PR DESCRIPTION
**Pull request by raduchis**
## Reasoning behind the pull request
- remove the chainRunType argument passed to multiple components
  
## Proposed changes
- inject the runTypeComponents needed for normal or sovereign 

## Testing procedure
- #allIn

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Review by sasurobert — DISMISSED** [review]

File: epochStart/bootstrap/process.go [hunk 1]
```diff
@@ -164,30 +164,30 @@ type baseDataInStorage struct {
...
+	NodeShuffler                    nodesCoordinator.NodesShuffler
+	RoundHandler                    epochStart.RoundHandler
+	ArgumentsParser                 process.ArgumentsParser
+	StatusHandler                   core.AppStatusHandler
+	HeaderIntegrityVerifier         process.HeaderIntegrityVerifier
+	DataSyncerCreator               types.ScheduledDataSyncerCreator
+	ScheduledSCRsStorer             storage.Storer
+	TrieSyncStatisticsProvider      common.SizeSyncStatisticsHandler
+	NodeProcessingMode              common.NodeProcessingMode
+	AdditionalStorageServiceCreator process.AdditionalStorageServiceCreator
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply epochStart/bootstrap/process.go:190]: Could stay as it is for now, though

File: epochStart/bootstrap/RequestHandlerFactoryStub.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package bootstrap
```
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/RequestHandlerFactoryStub.go:1]: there will be a cyclic import in test... tried to fix it cleaner, but didn't manage it.

File: epochStart/bootstrap/process.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/process.go:190]: would leave it like this for now

File: epochStart/bootstrap/storageProcess.go [hunk 3]
```diff
@@ -75,34 +81,18 @@ func NewStorageEpochStartBootstrap(args ArgsStorageEpochStartBootstrap) (*storag
...
 		chainID:                    args.CoreComponentsHolder.ChainID(),
 		timeToWaitForRequestedData: args.TimeToWaitForRequestedData,
-		chainRunType:               args.ChainRunType,
+		resolverRequestFactory:     args.ResolverRequestFactory,
 	}
 
 	return sesb, nil
 }
 
 func createEpochStartBootstrapper(args ArgsStorageEpochStartBootstrap) (EpochStartBootstrapper, error) {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/storageProcess.go:90]: fixed

File: epochStart/bootstrap/storageProcess_test.go [hunk 4]
```diff
@@ -33,31 +33,65 @@ func createMockStorageEpochStartBootstrapArgs(
 	coreMock *mock.CoreComponentsMock,
 	cryptoMock *mock.CryptoComponentsMock,
 ) ArgsStorageEpochStartBootstrap {
+	esbc, _ := NewEpochStartBootstrapperFactory()
```
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/storageProcess_test.go:36]: fixed

File: epochStart/bootstrap/storageProcess_test.go [hunk 5]
```diff
@@ -33,31 +33,65 @@ func createMockStorageEpochStartBootstrapArgs(
...
+	})
+	t.Run("nil ResolverRequestFactory should err", func(t *testing.T) {
+		t.Parallel()
+
+		coreComp, cryptoComp := createComponentsForEpochStart()
+		args := createMockStorageEpochStartBootstrapArgs(coreComp, cryptoComp)
+		args.ResolverRequestFactory = nil
+		sesb, err := NewStorageEpochStartBootstrap(args)
+		assert.True(t, check.IfNil(sesb))
+		assert.Error(t, mxErrors.ErrNilRequestHandlerCreator, err)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/storageProcess_test.go:93]: very good catch! it had to be ErrorIs. 
fixed in multiple locations

File: epochStart/bootstrap/storageProcess_test.go [hunk 6]
```diff
@@ -544,14 +555,12 @@ func TestStorageEpochStartBootstrap_applyCurrentShardIDOnMiniblocksCopy(t *testi
...
-	coreComp, cryptoComp := createComponentsForEpochStart()
-
 	t.Run("should create a main chain instance", func(t *testing.T) {
 		t.Parallel()
+		coreComp, cryptoComp := createComponentsForEpochStart()
 
 		args := createMockStorageEpochStartBootstrapArgs(coreComp, cryptoComp)
-		args.ChainRunType = common.ChainRunTypeRegular
-
+		args.ResolverRequestFactory, _ = requestHandlers.NewResolverRequestHandlerFactory()
```
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/storageProcess_test.go:563]: fixed

File: factory/api/apiResolverFactory_test.go [hunk 7]
```diff
@@ -107,7 +108,7 @@ func createMockArgs(t *testing.T) *api.ApiResolverArgs {
 		StatusComponents: &mainFactoryMocks.StatusComponentsStub{
 			ManagedPeersMonitorField: &testscommon.ManagedPeersMonitorStub{},
 		},
-		ChainRunType:       common.ChainRunTypeRegular,
+		ChainRunType: common.ChainRunTypeRegular,
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/api/apiResolverFactory_test.go:111]: will be removed in the next PR

File: testscommon/additionalStorageServiceFactoryMock.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package testscommon
+
+import "github.com/multiversx/mx-chain-go/dataRetriever"
+
+// AdditionalStorageServiceFactoryMock is a mock for AdditionalStorageServiceFactory
+type AdditionalStorageServiceFactoryMock struct {
+	WorkAsSovereign bool
+}
+
+// CreateAdditionalStorageUnits creates an additional storage units for a shard
```
[the remark this answers is not in the document]
Comment (raduchis) [reply testscommon/additionalStorageServiceFactoryMock.go:10]: fixed

File: cmd/sovereignnode/sovereignNodeRunner.go [hunk 9]
```diff
@@ -1497,6 +1505,33 @@ func (snr *sovereignNodeRunner) CreateManagedCryptoComponents(
 	return managedCryptoComponents, nil
 }
 
+// CreateManagedRunTypeComponents is the managed runType components factory
```
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/sovereignnode/sovereignNodeRunner.go:1508]: fixed

File: node/nodeRunner.go [hunk 10]
```diff
@@ -1534,6 +1547,26 @@ func (nr *nodeRunner) CreateManagedCryptoComponents(
 	return managedCryptoComponents, nil
 }
 
+// CreateManagedRunTypeComponents is the managed runType components factory
```
[the remark this answers is not in the document]
Comment (raduchis) [reply node/nodeRunner.go:1550]: fixed

File: factory/runType/interface.go [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package runType
+
+// RunTypeComponentsCreator is the interface for the runTypeComponentsCreator
+type RunTypeComponentsCreator interface {
+	Create() (*runTypeComponents, error)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/runType/interface.go:5]: I made the interface unexported so it is used just in this package.
I need to set a *runTypeComponent in the managedRunTypeComponents. I could return a ManagedRunTypeComponentesHolder and typecast before the asignment.

File: factory/processing/processComponentsHandler_test.go [hunk 12]
```diff
@@ -59,10 +59,11 @@ func testManagedProcessComponentsCreateShouldWork(t *testing.T, shardID uint32,
 	}
 
 	args := createMockProcessComponentsFactoryArgs()
+	if chainType == common.ChainRunTypeSovereign {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/processing/processComponentsHandler_test.go:62]: made it in the next PR.... but changed it here.

File: factory/processing/processComponents.go [hunk 13]
```diff
@@ -1422,24 +1346,7 @@ func (pcf *processComponentsFactory) createShardBlockTracker(argBaseTracker trac
...
-		pcf.blockTrackerCreator, err = track.NewSovereignBlockTrackerFactory(trackerFactory)
-		if err != nil {
-			return nil, err
-		}
-	default:
-		return nil, fmt.Errorf("%w type %v", errorsMx.ErrUnimplementedChainRunType, pcf.chainRunType)
-	}
-
-	return pcf.blockTrackerCreator.CreateBlockTracker(arguments)
+	return pcf.runTypeComponents.BlockTrackerCreator().CreateBlockTracker(arguments)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/processing/processComponents.go:1349]: removed

File: factory/processing/processComponents.go [hunk 14]
```diff
@@ -923,25 +865,7 @@ func (pcf *processComponentsFactory) newEpochStartTrigger(requestHandler epochSt
 }
 
 func (pcf *processComponentsFactory) createHeaderValidator(argsHeaderValidator block.ArgsHeaderValidator) (process.HeaderConstructionValidator, error) {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/processing/processComponents.go:867]: fixed

File: factory/processing/processComponents.go [hunk 15]
```diff
@@ -833,25 +793,7 @@ func (pcf *processComponentsFactory) newValidatorStatisticsProcessor() (process.
 }
 
 func (pcf *processComponentsFactory) createValidatorStatisticsProcessor(args peer.ArgValidatorStatisticsProcessor) (process.ValidatorStatisticsProcessor, error) {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/processing/processComponents.go:795]: fixed

File: factory/processing/blockProcessorCreator.go [hunk 16]
```diff
@@ -451,50 +450,13 @@ func (pcf *processComponentsFactory) newShardBlockProcessor(
 func (pcf *processComponentsFactory) createTransactionCoordinator(
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/processing/blockProcessorCreator.go:450]: fixed

File: factory/data/dataComponentsHandler_test.go [hunk 17]
```diff
@@ -26,9 +26,10 @@ func TestNewManagedDataComponents(t *testing.T) {
 	t.Run("should work", func(t *testing.T) {
 		t.Parallel()
 
+		runTypeComponents := componentsMock.GetRunTypeComponents()
 		coreComponents := componentsMock.GetCoreComponents()
 		shardCoordinator := mock.NewMultiShardsCoordinatorMock(2)
-		args := componentsMock.GetDataArgs(coreComponents, shardCoordinator)
+		args := componentsMock.GetDataArgs(coreComponents, runTypeComponents, shardCoordinator)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/data/dataComponentsHandler_test.go:32]: fixed, yes it is cleaner

File: factory/data/dataComponents.go [hunk 18]
```diff
@@ -14,38 +14,41 @@ import (
...
+	ShardCoordinator                sharding.Coordinator
+	Core                            factory.CoreComponentsHolder
+	StatusCore                      factory.StatusCoreComponentsHolder
+	Crypto                          factory.CryptoComponentsHolder
+	FlagsConfigs                    config.ContextFlagsConfig
+	CurrentEpoch                    uint32
+	CreateTrieEpochRootHashStorer   bool
+	NodeProcessingMode              common.NodeProcessingMode
+	AdditionalStorageServiceCreator process.AdditionalStorageServiceCreator
+	ChainRunType                    common.ChainRunType
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/data/dataComponents.go:36]: will come in the next PR... I stopped at ~1400 lines of code.

File: epochStart/bootstrap/storageProcess_test.go [hunk 19]
```diff
@@ -121,6 +127,10 @@ func TestStorageEpochStartBootstrap_BootstrapStartInEpochNotEnabled(t *testing.T
 			return storage.LatestDataFromStorage{}, err
 		},
 	}
+
+	esbc, _ := NewEpochStartBootstrapperFactory()
+
+	args.EpochStartBootstrapperCreator = esbc
```
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/storageProcess_test.go:133]: fixed

File: factory/bootstrap/bootstrapComponents_test.go [hunk 20]
```diff
@@ -254,7 +286,7 @@ func TestBootstrapComponentsFactory_CreateEpochStartBootstrapperShouldWork(t *te
 		t.Parallel()
 
 		args := componentsMock.GetBootStrapFactoryArgs()
-		args.ChainRunType = common.ChainRunTypeRegular
+		args.RunTypeComponents = componentsMock.GetRunTypeComponents()
```
[the remark this answers is not in the document]
Comment (raduchis) [reply factory/bootstrap/bootstrapComponents_test.go:289]: fixed, added check for epochStartBootstrapper and sovereignEpochStartBootstrapper

File: cmd/sovereignnode/sovereignNodeRunner.go [hunk 21]
```diff
@@ -1497,6 +1505,33 @@ func (snr *sovereignNodeRunner) CreateManagedCryptoComponents(
...
+func (snr *sovereignNodeRunner) CreateManagedRunTypeComponents() (mainFactory.RunTypeComponentsHandler, error) {
+	runTypeComponentsFactory, err := runType.NewRunTypeComponentsFactory()
+	if err != nil {
+		return nil, fmt.Errorf("newRunTypeComponentsFactory failed: %w", err)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/sovereignnode/sovereignNodeRunner.go]: committed

**Review by mariusmihaic — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**